internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -183,11 +187,18 @@ type CompositionRequest struct {
 	Environment *Environment
 }
 
+// FunctionCondition represents a condition returned by a Composition Function
+type FunctionCondition struct {
+	conditions.Condition
+	Target v1beta1.Target
+}
+
 // A CompositionResult is the result of the composition process.
 type CompositionResult struct {
 	Composed          []ComposedResource
 	ConnectionDetails managed.ConnectionDetails
-	Events            []event.Event
+	Events            map[v1beta1.Target][]event.Event
+	Conditions        map[xpv1.ConditionType]FunctionCondition
```

Comment:
Now that result handling is a little more complicated, I'm wondering whether it makes sense to parse them out into event and conditions in `composition_functions.go`, or here. 🤔 

It would be ideal not to leak details of the function implementation up at this level though. Thinking out loud, could we:

* Add a new type like `CompositionEvent` or something that is pretty much the same as `Result`
* Have both `Composer` implementations return an array of `CompositionEvent`

Parse the `[]CompositionEvent` out into events and status conditions up at this level?

I expect some of the challenges with that approach would be:

* We don't want claim events to talk about functions and steps
* We'd still want fatal events to stop the `Composer` implementation